server/cluster/cluster.go

```diff
@@ -665,6 +664,70 @@ func (c *RaftCluster) processRegionHeartbeat(region *core.RegionInfo) error {
 		time.Sleep(500 * time.Millisecond)
 	})
 
+	return c.saveRegion(isNew, saveKV, saveCache, needSync, region, origin)
+}
+
+// processRegionHeartbeatResponse updates the region information thought region split.
+// It will update peer that not exist in the cache, and add new peers that join in.
+func (c *RaftCluster) processRegionSpiltReport(regions []*metapb.Region) (errs []error) {
+	if err := checkSplitRegions(regions); err != nil {
+		return []error{err}
+	}
+	total := len(regions) - 1
+	c.RLock()
+	coreCluster := c.core
+	c.RUnlock()
+	leaderStoreID := uint64(0)
+	if r := coreCluster.GetRegion(regions[0].GetId()); r != nil {
+		leaderStoreID = r.GetLeader().GetStoreId()
+	}
+	for i := range regions {
+		// It should update rightmost because it can override by left.
+		index := total - i
+		v := regions[index]
+		if len(v.GetPeers()) == 0 {
+			errs = append(errs, errors.New(fmt.Sprintf("region:%d has no peer", v.GetId())))
+			continue
+		}
+		// region split initiator store will be leader with a high probability
+		leader := v.Peers[0]
+		if leaderStoreID > 0 {
+			for _, peer := range v.GetPeers() {
+				if peer.GetStoreId() == leaderStoreID {
+					leader = peer
+					break
+				}
+			}
+		}
+		region := core.NewRegionInfo(v, leader)
+		origin, perr := coreCluster.PreCheckPutRegion(region)
+		if perr != nil {
+			errs = append(errs, perr)
+			continue
+		}
+		update, isNew, needSync := false, false, false
+		// The region that is not the rightmost element should update if origin is nil.
+		// The rightmost region should update if epoch greater than origin.
+		if origin == nil {
+			update = true
+			isNew = true
+			needSync = true
+		} else if index == total {
+			r, o := region.GetRegionEpoch(), origin.GetRegionEpoch()
+			if r.GetConfVer() > o.GetConfVer() {
+				update = true
+			}
+		}
+		if update {
+			if err := c.saveRegion(isNew, true, true, needSync, region, origin); err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+	return errs
+}
+
+func (c *RaftCluster) saveRegion(isNew, saveKV, saveCache, needSync bool, region *core.RegionInfo, origin *core.RegionInfo) error {
```

Comment:
Maybe it is necessary to put `isNew`, `saveKV`, `saveCache`, `needSync` in a `Struct`.